networkit/cpp/generators/ClusteredRandomGraphGenerator.cpp

```diff
@@ -3,55 +3,114 @@
+
+    zeta.setUpperBound(k);
 }
 
 Graph ClusteredRandomGraphGenerator::generate() {
-    assert(pin >= pout);
+    GraphBuilder graphBuilder(n);
 
-    Graph G(n);
-    // assign nodes evenly to clusters
-    Partition zeta(n);
-    zeta.setUpperBound(k);
-    G.forNodes([&](node v){
-        index c = Aux::Random::integer(k-1);
-        zeta.addToSubset(c, v);
-    });
-
-    assert (zeta.numberOfSubsets() == k);
-
-    G.forNodePairs([&](node u, node v){
-        if (zeta.subsetOf(u) == zeta.subsetOf(v)) {
-            if (Aux::Random::probability() <= pin) {
-                G.addEdge(u, v);
-            }
-        } else {
-            if (Aux::Random::probability() <= pout) {
-                G.addEdge(u, v);
-            }
+    // Sequence of all nodes grouped by cluster
+    std::vector<node> clustersSequence;
+
+    // Index where each cluster starts in `clustersSequence`
+    std::vector<size_t> clusterStart(k);
+
+    {
+        std::uniform_int_distribution<index> clusterDistr{0, k - 1};
+
+        // Nodes grouped by cluster
+        std::vector<std::vector<node>> clusters(k);
+
+        // Assign nodes evenly to clusters
+        graphBuilder.parallelForNodes([&zeta = zeta, &clusters, &clusterDistr](node u) {
+            const index cluster = clusterDistr(Aux::Random::getURNG());
+            zeta.addToSubset(cluster, u);
+#pragma omp critical
+            clusters[cluster].push_back(u);
+        });
+
+        for (index i = 1; i < k; ++i)
+            clusterStart[i] = clusterStart[i - 1] + clusters[i - 1].size();
+
+        for (auto &&cluster : clusters)
+            clustersSequence.insert(clustersSequence.end(), cluster.begin(), cluster.end());
+    }
+
+    // Add all the half-out edges from a node u to all the other nodes in the given sequence
+    const auto addHalfOutEdges = [&graphBuilder, &clustersSequence](node u, index start, index end,
+                                                                    double p) -> void {
+        if (end - start <= 1 || p == 0) // No edges to add
+            return;
+
+        if (p == 1) { // Add all edges in the interval
+            for (index i = start + 1; i < end; ++i)
+                graphBuilder.addHalfOutEdge(u, clustersSequence[i]);
+            return;
         }
-    });
 
-    this->zeta = std::move(zeta);
+        std::geometric_distribution<index> geom(p);
+        do {
+            const index next = start + geom(Aux::Random::getURNG()) + 1;
+            if (next < end) {
+                graphBuilder.addHalfEdge(u, clustersSequence[next]);
+                start = next;
+            } else
+                return;
+        } while (true);
+    };
 
-    G.shrinkToFit();
-    return G;
+#pragma omp parallel for schedule(guided)
+    for (omp_index i = 0; i < static_cast<omp_index>(n); ++i) {
+        const node u = clustersSequence[i];
+        const index clusterIdx = zeta.subsetOf(u);
+        const index start = clusterStart[clusterIdx];
```

Comment:
Shouldn't start be `i`? From what I understood, otherwise, there will be twice as many intra-cluster edges added as desired to all nodes but the first node of the cluster and in particular also self-loops. Adding some documentation that documents the assumptions made here and in particular in `addHalfOutEdges` could help here (e.g., `addHalfOutEdges` seems to assume that the position of `u` is `start`, which seems wrong for inter-cluster edges).

Reply:
That's true, thanks. I also added further documentation.